File: 300_Longest_Increasing_Subsequence.py
class Solution:
    def lengthOfLIS(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        n = len(nums)
        if n < 2:
            return n
        dp = [1 for __ in range(n)]
        for i in (range(1, n)):
            for j in (range(0, i)):
                if nums[i] > nums[j]:
                    dp[i] = max(dp[j] + 1, dp[i])
        return max(dp)


    def lengthOfLIS(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        tails = [0] * len(nums)
        size = 0
        for x in nums:
            i, j = 0, size
            while i != j:
                m = int((i + j) / 2)
                if  tails[m] < x:
                    i = m + 1
                else:
                    j = m
            tails[i] = x
            size = max(i + 1, size)
        return size

File: 300_Longest_Increasing_Subsequence.cpp
class Solution {
public:
    int lengthOfLIS(vector<int>& nums) {
        int size = nums.size();
        if (size < 2) return size;
        int maximum = 0;
        vector<int> dp (size, 1);
        for (int i = 1; i < size; ++i) {
            for (int j = 0; j < i; ++j) {
                if (nums[i] > nums[j]) {
                    dp[i] = max(dp[j] + 1, dp[i]);
                }
            }
        }
        
        for (int i = 0; i < size; ++i) {
            maximum = max(maximum, dp[i]);
        }
        return maximum;
    }
};


class Solution {
public:
    int lengthOfLIS(vector<int>& nums) {
        vector<int> res;
        for (int num : nums) {
            auto it = lower_bound(res.begin(), res.end(), num);
            if (it == res.end()) res.push_back(num);
            else *it = num;
        }
        
        return res.size();
    }
};
